cpp/src/arrow/types/primitive.h

```diff
@@ -54,30 +54,54 @@ class PrimitiveArray : public Array {
-  T Value(int i) const {                                        \
-    return raw_data()[i];                                       \
-  }                                                             \
+#define NUMERIC_ARRAY_DECL(NAME, TypeClass, T)                              \
+class NAME : public PrimitiveArray {                                        \
+ public:                                                                    \
+  using value_type = T;                                                     \
+  using PrimitiveArray::PrimitiveArray;                                     \
+                                                                            \
+  NAME(int32_t length, const std::shared_ptr<Buffer>& data,                 \
+      int32_t null_count = 0,                                               \
+      const std::shared_ptr<Buffer>& null_bitmap = nullptr) :               \
+      PrimitiveArray(std::make_shared<TypeClass>(), length,                 \
+          data, null_count, null_bitmap) {}                                 \
+                                                                            \
+  bool EqualsExact(const NAME& other) const {                               \
+    return PrimitiveArray::EqualsExact(                                     \
+        *static_cast<const PrimitiveArray*>(&other));                       \
+  }                                                                         \
+                                                                            \
+  std::shared_ptr<NAME> slice(int32_t start) const{                         \
+    return slice(start, length_);                                           \
+  }                                                                         \
+                                                                            \
+  std::shared_ptr<NAME> slice(int32_t start, int32_t length) const{         \
+    int32_t uFrom = start;                                                  \
+    int32_t uLength = length;                                               \
+    while(uFrom < 0 || uLength < 0) {                                       \
+      uFrom = start >= 0 ? start : length_ + start;                         \
+      uLength = length >= 0 ? length : length_ + length;                    \
+    }                                                                       \
+                                                                            \
+    if (uFrom > uLength || uFrom > length_) return nullptr;                 \
+    if (uLength > (length_ - start)) uLength = length_ - start;             \
+                                                                            \
+    auto parent_data = data_;                                               \
+    auto sliced_buf = std::make_shared<Buffer>(parent_data,                 \
+           int64_t(uFrom * sizeof(T)), int64_t(uLength * sizeof(T)));       \
```

Comment:
These casts are not needed. If a cast is required, use `static_cast<int64_t>(...)`